### ISSUE TITLE
Run failed with `wasi_snapshot_preview1.wasm`

### ISSUE THREAD

**Issue by fuxiaohei** [bytecodealliance/preview2-prototyping#78]
I create a demo repo to show this case. https://github.com/fuxiaohei/wasmtime-component-wasi-preview-test

wit code:

```wit
interface simple {
  print: func()
}

default world test-world {
  export exports: self.simple
}
```
generate exports code in https://github.com/fuxiaohei/wasmtime-component-wasi-preview-test/blob/main/wasm_lib/src/lib.rs with latest `wit-bindgen-guest-rust = { git = "https://github.com/bytecodealliance/wit-bindgen" }`

then I build wasm32-wasi target and convert to component with same code in `host tests` in this repo and current `wasi_snapshot_preview1.wasm` in release page.
https://github.com/fuxiaohei/wasmtime-component-wasi-preview-test/blob/main/src/main.rs#L5

but it runs failed and shows error msg:

```
Convert wasm module to component success, target/wasm32-wasi/release/wasm_lib.component.wasm
Run component: target/wasm32-wasi/release/wasm_lib.component.wasm
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: error while executing at wasm backtrace:
    0: 0x204d9a - wit-component:shim!adapt-wasi_snapshot_preview1-environ_sizes_get
    1: 0x8ef5 - <unknown>!__wasi_environ_sizes_get
    2: 0x8f5c - <unknown>!__wasilibc_initialize_environ
    3: 0x8ffc - <unknown>!__wasilibc_initialize_environ_eagerly
    4:  0x570 - <unknown>!__wasm_call_ctors
    5: 0xda10 - <unknown>!cabi_realloc.command_export
    6: 0x204493 - wit-component:adapter:wasi_snapshot_preview1!initialize_stack_pointer
note: using the `WASMTIME_BACKTRACE_DETAILS=1` environment variable may show more debugging information

Caused by:
    wasm trap: uninitialized element', src/main.rs:51:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

You can follow the [readme](https://github.com/fuxiaohei/wasmtime-component-wasi-preview-test#readme) and reproduce this case. 
It can reproduce in actions: https://github.com/fuxiaohei/wasmtime-component-wasi-preview-test/actions/runs/4089812259.

maybe I miss something details to convert module to component in wasi case.

**Comment by pchickey** [comment]
This is expected, we dont have a non-command implementation of the environment yet, I'll work on fixing that soon by changing the way environments work in commands.

**Comment by alexcrichton** [comment]
Hm actually wait I don't think this is related to environments and comments. @dicej this is a problem with the recent changes to use `cabi_realloc`.

The problem here is:

* The main module is instantiated with "shims" that all trap with `uninitialized element`, the error message here, if an import is called before it's ready.
* The adapter module is then instantiated with the main module as its imports
* The adapter module's `start` function runs, which calls the main module's `cabi_realloc`
* The main module's `cabi_realloc` calls `__wasm_call_ctors`
* The `__wasm_call_ctors` function calls the WASI `environ_sizes_get` function.
* This function originally started out as an adapter, so it was a "trapping import" at the start. This gets filled in after the adapter is instantiated, but it's not filled in yet, so `environ_sizes_get` traps.

The tl;dr; here is that due to all exports invoking constructors, including the export of malloc, this means that using the main module's `cabi_realloc` is not feasible. Unless there's a way to export something from the main module without getting it hooked up into ctors/dtors/etc which I'm not sure how to do. cc @sunfishcode on this as well

**Comment by sunfishcode** [comment]
In Rust 1.67 and wasi-sdk-17, ctors/dtors should no longer get hooked up to exports other than `_start`, which may help with this.

**Comment by alexcrichton** [comment]
Unfortunately I don't think that's sufficient. Notably the `cdylib` crate type, used in @fuxiaohei's example, still gets the ctors/dtors treatment along with `*.command_export` naming, even on the latest nightly.

**Comment by dicej** [comment]
This is specifically the stack allocation that's happening in the `start` function, right?  Whereas the Preview 2 adapter lazily initializes `State` and thus shouldn't call `cabi_realloc` until after the `start` function has run?  Could we do the same thing for stack allocation, i.e. instrument every adapter export with a check to see if the stack pointer has been initialized and then allocate it if necessary?

**Comment by dicej** [comment]
I just did a quick call with Alex to discuss this.  The lazy stack allocation strategy I proposed above _could_ work (at least for certain scenarios) -- I'll do some more research and report back here.

**Comment by dicej** [comment]
Update on this: Alex and I came up with a strategy to address this which involves (A) lazily allocating the stack, (B) short-circuiting reentrant calls into the adapter, and (C) ensuring that unexpected use of the stack prior to allocation will trap.  I'll be working on this today and will post a PR ASAP.

**Comment by dicej** [comment]
I have a fix for this, and it seems to resolve @fuxiaohei's test case.  I need to rebase and test some other scenarios before opening PRs; meanwhile, you can check out the code here:
https://github.com/dicej/wasm-tools/commit/f95d4a151818dc40a632ad92ea7e14fa1e120ae3
https://github.com/dicej/preview1.wasm/commit/2a57c696cd0dcf03f43316d6362fdfac6b3e41db
